### ISSUE TITLE
The Fréchet mean on SPD matrices doesn't return a symmetric matrix

### ISSUE THREAD

**Issue by miguelgondu** [geomstats/geomstats#1547]
Hi everyone,

I'm currently using `geomstats` with PyTorch as a backend. I'm currently replicating the results of [_Wrapped Gaussian Process Regression on Riemannian Manifolds_](https://openaccess.thecvf.com/content_cvpr_2018/papers/Mallasto_Wrapped_Gaussian_Process_CVPR_2018_paper.pdf). The author provides the data for the SPD matrices experiment in [this link](https://drive.google.com/uc?export=download&id=1psVU_kw86gqYLD9TlyLGvFba61dlWayB).

When using the built-in `FrechetMean` object, the result of the estimation is not symmetric. Here's a minimal example:

```python
from scipy.io import loadmat

import torch

from geomstats.geometry.spd_matrices import SPDMetricAffine
from geomstats.learning.frechet_mean import FrechetMean

data = loadmat("DTI_data.mat")
targets = torch.from_numpy(data["pop_orig"].T)

metric = SPDMetricAffine(3)
frechet_mean = FrechetMean(metric)
frechet_mean.fit(targets)
mean_estimate = frechet_mean.estimate_

print(mean_estimate)
print(mean_estimate.T == mean_estimate)
assert (mean_estimate.T == mean_estimate).all()
```

To download the data, follow [this link](https://drive.google.com/uc?export=download&id=1psVU_kw86gqYLD9TlyLGvFba61dlWayB). This prints:

```
tensor([[ 9.4671e-05,  3.7080e-06,  5.8562e-06],
        [ 3.7080e-06,  7.4874e-05, -1.0567e-09],
        [ 5.8562e-06, -1.0634e-09,  9.1638e-05]])
tensor([[ True, False, False],
        [False,  True, False],
        [False, False,  True]])
```

The floats are close, of course, but they are far away enough to cause stability issues in e.g. a Wrapped Gaussian Process implementation.

One solution would be adding a projection, but I wonder if we should burden the user with this, or the library itself. What do you think?

Details on my set-up:
- OS: Ubuntu 20.04.
- Python: `3.9.12`.
- geomstats version: `2.5.0`.
- Using torch as a backend: torch version: `1.11`.

**Comment by ninamiolane** [comment]
Thank you for your interest! @tomszwagier @arthurPignet any idea there?

A possibility would be to enforce float64 with Pytorch (which is not the default, the default is float32) which could solve the numerical issue.

**Comment by miguelgondu** [comment]
Thanks for the quick answer! About that: something weird happens when e.g. I run `torch.set_default_dtype(torch.float64)`. It's as if geomstats doesn't get the memo, and it complains when I pass double precision tensors.

I can be more specific and provide an example if you want.

I agree that enforcing double precision would make sense, given that these algorithms are sometimes very unstable, numerically. What do the rest think?

**Comment by tomszwagier** [comment]
Hello @miguelgondu,

First of all, it might be in your interest to know that @arthurPignet implemented WGPR on Riemannian manifolds in `geomstats`.
You can find the code here: https://github.com/geomstats/geomstats/blob/master/geomstats/learning/wrapped_gaussian_process.py
Then we submitted a Notebook with some experiments pretty close from the original paper to "ICLR Computational Geometry & Topology Challenge 2022".
The submission is here: https://github.com/geomstats/challenge-iclr-2022/blob/main/2nd-prize__szwagier__pignet__Wrapped-Gaussian-Process-Regression-on-Manifolds/Wrapped_Gaussian_Process_Regression_on_Riemannian_Manifolds.ipynb

I don't know when you said that you were trying to replicate the paper results if you meant that you were implenting WGPR from scratch in `geomstats` or if you were using our implementation; so in the first case, it might be interesting for you to compare the implementations. @lpereira95 made some recent changes on the implementation, so it might be important to keep up to date with the latest version (moreover we were discussing with @lpereira95 recently about trying to replicate the paper results so I am adding you to the loop because this might be interesting to mix the works).

Second, it is nice that you managed to load the `.mat` file with DTI data coming from the paper, because I tried that when working on the submission and suprisingly couldn't load it on Python with the same code as yours (`loadmat("DTI_data.mat")`). But I just tried your code now and it works so it might have been an issue of package version.

Finally, to come back to the original question, I don't remember checking that the Fréchet Mean was a Symmetric Matrix when working on the submission. Maybe @arthurPignet you did when implementing the tests? Anyway I am not very familiar with those numerical stability issues, but enforcing double precision totally makes sense in that case to me, rather than projecting.

**Comment by miguelgondu** [comment]
Hi Tom,

Thanks for the resources!, and congrats on getting the second place for the challenge. The implementation you and @arthurPignet did is pretty neat, and I have toyed around with it. I have some questions about how you deal with the Euclidean GPR on the tangent space, but we could take them offline on geomstats' Slack.

If we all agree on the solution of enforcing `float64`s for PyTorch, I could try to implement that change and send it as a PR. Moreover, why not move one of the examples in your submission to the `examples` folder? I usually check them when trying to understand how to train a learning algorithm in `geomstats`.

**Comment by lpereira95** [comment]
Hi @miguelgondu. Thanks a lot for your suggestions and issue creation (I'm following them but unfortunately haven't had the time to help... hopefully soon).

For this case, I disagree we have a problem (even using the default pytorch precision). Have you had stability issues when running e.g. WGPs? If yes, can you open a new PR with a minimal working example?

I say we do not really have a problem due to 2 reasons. First, if we do:

```
spd_mat = SPDMatrices(3)

spd_mat.belongs(mean_estimate)
```

the result is `True`. Which means that for the set `gs.atol` (which is 1e-6), the method does what is expected.

Second, because:

```
mean_estimate - mean_estimate.T
```

outputs a matrix which highest value is 1e-12, which is much smaller than  the tolerances set.

Lastly, we can also control the "quality" of `FrechetMean` output by setting `epsilon` and `max_iter`. Notice that by default `epsilon=1e-4`, which means the method is doing what is it asked for.

I'll try to understand what's going on with `torch.set_default_dtype`...

Out of curiosity, do you see any advantage of using `pytorch` instead of `numpy` here as backend?


P.S. @tomszwagier, the last release of `geomstats` does not contain my changes to WGPs, which mean @miguelgondu is using the version implemented by @arthurPignet.

**Comment by tomszwagier** [comment]
Hi,
It may be a great idea to move some experiments in the examples folder indeed @miguelgondu.
Regarding using `pytorch `instead of `numpy`, @lpereira95, I could add that when we made the experiments for the submission, the `GeodesicRegression` wasn't working in `numpy` backend if I remember well, but only in `pytorch` and `autograd` if I remember well. But I'll let @miguelgondu answer why in is case.
Ok @lpereira95 I didn't know, thank you.

**Comment by arthurPignet** [comment]
hello all,

@tomszwagier > Finally, to come back to the original question, I don't remember checking that the Fréchet Mean was a Symmetric Matrix when working on the submission. Maybe @arthurPignet you did when implementing the tests? 

In the code of `wgpr` the prior is used to compute basepoints, which are then used in log/exp of the given metric. I think that those functions would raise an error if the used basepoint doesn't belongs to the manifold. (I am not 100% sure of that, but I remember running into this kind of error when prototyping the code)

@lpereira95 > Out of curiosity, do you see any advantage of using pytorch instead of numpy here as backend?

As @tomszwagier  said, we needed the autograd stuffs to do a `GeodesicRegression`, and use it as prior. However, as the `wgpr` strongly relies on `sklearn`, I think using `numpy` avoid passing constantly from `torch.Tensors` to `numpy.arrays`, and avoid a layer of complexity. All the more when a GPU is used as default device. 

@miguelgondu > I have some questions about how you deal with the Euclidean GPR on the tangent space, but we could take them offline on geomstats' Slack.

I would be a pleasure to discuss that with you. Don't hesitate to reach out

**Comment by lpereira95** [comment]
@arthurPignet your point on tensor conversion when the device is gpu is exactly what is in my head. That's why for this specific case I would probably have used `autograd` (we will also add `jax` soon, which would probably be a better choice here) - also because in cpu `numpy` is probably faster. Nevertheless I would like to see a performance comparison: maybe for a sufficiently large problem size torch in gpu is faster... (in the future we may have to address this better gpu-cpu conversion better -> maybe we would have to provide our own implementation of GPs or rely on an implementation that runs on GPU - e.g. `gpytorch`).
